src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs

```diff
@@ -32,6 +32,9 @@ public async Task Invoke(MessageContext messageContext, CancellationToken cancel
                 var rootContext = new RootContext(childScope.ServiceProvider, messageOperations, pipelineCache, cancellationToken);
                 rootContext.Extensions.Merge(messageContext.Extensions);
 
+                // We set the outer context to allow access for any settings we need to propagate up
+                rootContext.Extensions.Set("SharedContext", messageContext.Extensions);
```

Comment:
would it be useful to have some kind of constant?

we should also use a different (and prefixed) name that has a smaller risk of collision with user-selected keys.